python/pyspark/ml/torch/distributor.py

```diff
@@ -0,0 +1,491 @@
+
+    Parameters
+    ----------
+    cmd : list[str]
+        The users's input command split by the space characeter
+    _prctl : bool, optional
+        (test only) Use prctl to signal the command process upon parent death, by default True
+    redirect_to_stdout : bool, optional
+        whether to redirect all stdout/stderr logs to current stdout, by default True
+    TODO(SPARK-41593): Will update with additional parameters for logging to driver
+
+    Raises
+    ------
+    RuntimeError
+        If the child executed command fails
+    """
+    _TAIL_LINES_TO_KEEP = 100
+
+    def sigterm_on_parent_death() -> None:
+        """
+        Uses prctl to automatically send SIGTERM to the command process when its parent is dead.
+        This handles the case when the parent is a PySpark worker process.
+        If a user cancels the PySpark job, the worker process gets killed, regardless of
+        PySpark daemon and worker reuse settings.
+        NOTE: This doesn't work on windows or POSIX.
+        Future reference:
+        - MacOS/POSIX: https://stackoverflow.com/a/2035683
+        - Windows: https://stackoverflow.com/a/23587108
+        """
+        if _prctl:
+            try:
+                libc = ctypes.CDLL("libc.so.6")
+                # Set the parent process death signal of the command process to SIGTERM.
+                libc.prctl(1, signal.SIGTERM)
+            except OSError:
+                pass
+
+    task = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        stdin=subprocess.PIPE,
+        env=os.environ,  # preserve the environ to support nested MPI jobs
+        preexec_fn=sigterm_on_parent_death,
+    )
+    task.stdin.close()  # type: ignore
+    tail: collections.deque = collections.deque(maxlen=_TAIL_LINES_TO_KEEP)
+    try:
+        for line in task.stdout:  # type: ignore
+            decoded = line.decode()
+            tail.append(decoded)
+            if redirect_to_stdout:
+                sys.stdout.write(decoded)
+        task.wait()
+    finally:
+        if task.poll() is None:
+            try:
+                task.terminate()  # SIGTERM
+                time.sleep(0.5)
+                if task.poll() is None:
+                    task.kill()  # SIGKILL
+            except OSError:
+                pass
+    if task.returncode != os.EX_OK:
+        if len(tail) == _TAIL_LINES_TO_KEEP:
+            last_n_msg = f"last {_TAIL_LINES_TO_KEEP} lines of the task output are"
+        else:
+            last_n_msg = "task output is"
+        task_output = "".join(tail)
+        raise RuntimeError(
+            f"Command {cmd} failed with return code {task.returncode}."
+            f"The {last_n_msg} included below: {task_output}"
+        )
+
+
+class Distributor:
+    def __init__(
+        self,
+        num_processes: int = 1,
+        local_mode: bool = True,
+        use_gpu: bool = True,
+        spark: Optional[SparkSession] = None,
```

Comment:
This param is meaningless, because we can only create one spark session or get active spark session.

So I suggest remove this param, and we can get spark session by:
`SparkSession.getActiveSession()`, if active spark session is None, raise error and ask user to create spark session first.

I don't recommend internally get spark session by `SparkSession.builder.getOrCreate()`, if no active spark session, this line will create a local mode spark session with default config, which might be not what user want.